rclcpp/test/rclcpp/executors/test_events_executor.cpp

```diff
@@ -479,6 +479,73 @@ TEST_F(TestEventsExecutor, destroy_entities)
   spinner.join();
 }
 
+// This test verifies that the add_node operation is robust wrt race conditions.
+// The initial implementation of the events-executor contained a bug where the executor
+// would end up in an inconsistent state and stop processing interrupt/shutdown notifications.
+// Manually adding a node to the executor results in a) producing a notify waitable event
+// and b) refreshing the executor collections.
+// The inconsistent state would happen if the event was processed before the collections were
+// finished to be refreshed: the executor would pick up the event but be unable to process it.
+// This would leave the `notify_waitable_event_pushed_` flag to true, preventing additional
+// notify waitable events to be pushed.
+// The behavior is observable only under heavy load, so this test spawns several worker
+// threads. Due to the nature of the bug, this test may still succeed even if the
+// bug is present. However repeated runs will show its flakiness nature and indicate
+// an eventual regression.
+TEST_F(TestEventsExecutor, testRaceConditionAddNode)
+{
+  // rmw_connextdds doesn't support events-executor
+  if (std::string(rmw_get_implementation_identifier()).find("rmw_connextdds") == 0) {
+    GTEST_SKIP();
+  }
+
+  // Spawn some threads to do some heavy work
+  std::atomic<bool> should_cancel = false;
+  std::vector<std::thread> stress_threads;
+  for (size_t i = 0; i < 5 * std::thread::hardware_concurrency(); i++) {
```

Comment:
with these stress threads, I'm able to reproduce the bug ~75% of the times on my laptop.

Reply:
Just for reference, what is the value of hardware_concurrency on your system?

Reply:
It's 16.

Note that the failure is reproducible also without these worker threads, but it takes many more trials (IIRC the failure rate was around 1%)